iobuf: optimize comparison more

rockwotj opened the pull request:
Turns out if the first byte is different memcmp is slower than a byte
for byte comparison. So what we do is optimize small iobufs and change
how we do the comparison, instead we check. This is a tradeoff we can
see in the bench for the best/worse case of how many bytes we compare


## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
iobuf: optimize comparison more

Turns out if the first byte is different memcmp is slower than a byte
for byte comparison. So what we do is optimize small iobufs and change
how we do the comparison, instead we check. This is a tradeoff we can
see in the bench for the best/worse case of how many bytes we compare

rockwotj commented:
```
test                        iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.cmp_bench_0000         333636000     2.711ns     0.001ns     2.710ns     2.712ns       0.000       0.000        39.2        10.0
iobuf.cmp_bench_0001         275198000     3.294ns     0.002ns     3.284ns     3.298ns       0.000       0.000        53.7        12.4
iobuf.cmp_bench_0002         278894000     3.241ns     0.000ns     3.241ns     3.244ns       0.000       0.000        49.7        12.2
iobuf.cmp_bench_0003         298818000     3.003ns     0.004ns     2.998ns     3.041ns       0.000       0.000        49.7        11.2
iobuf.cmp_bench_0004         299374000     2.996ns     0.001ns     2.994ns     2.997ns       0.000       0.000        49.7        11.2
iobuf.cmp_bench_0016         291622000     2.999ns     0.001ns     2.995ns     3.000ns       0.000       0.000        49.7        11.2
iobuf.cmp_bench_0064         176804000     5.098ns     0.001ns     5.092ns     5.100ns       0.000       0.000        58.7        19.7
iobuf.cmp_bench_0128         169316000     5.101ns     0.002ns     5.097ns     5.105ns       0.000       0.000        58.7        19.7
iobuf.cmp_bench_0256         158622000     5.094ns     0.003ns     5.090ns     5.097ns       0.000       0.000        58.7        19.7
iobuf.cmp_bench_0512         138610000     5.104ns     0.003ns     5.096ns     5.107ns       0.000       0.000        58.7        19.7
iobuf.cmp_bench_1024         101416000     5.824ns     0.002ns     5.822ns     5.826ns       0.000       0.000        67.2        22.7
iobuf.cmp_bench_2048          71142000     5.823ns     0.000ns     5.823ns     5.824ns       0.000       0.000        67.2        22.7
iobuf.cmp_bench_0000_same    326342000     2.763ns     0.003ns     2.760ns     2.767ns       0.000       0.000        39.2        10.2
iobuf.cmp_bench_0001_same    274102000     3.309ns     0.002ns     3.293ns     3.311ns       0.000       0.000        53.7        12.4
iobuf.cmp_bench_0002_same    263956000     3.450ns     0.001ns     3.448ns     3.452ns       0.000       0.000        59.7        12.9
iobuf.cmp_bench_0003_same    244180000     3.749ns     0.002ns     3.747ns     3.752ns       0.000       0.000        65.7        14.2
iobuf.cmp_bench_0004_same    223318000     4.134ns     0.001ns     4.129ns     4.135ns       0.000       0.000        71.7        15.7
iobuf.cmp_bench_0016_same    113770000     8.108ns     0.032ns     8.076ns     8.193ns       0.000       0.000       143.7        32.1
iobuf.cmp_bench_0064_same    176214000     5.108ns     0.001ns     5.107ns     5.109ns       0.000       0.000        64.7        19.7
iobuf.cmp_bench_0128_same    153394000     5.778ns     0.002ns     5.775ns     5.780ns       0.000       0.000        76.7        22.4
iobuf.cmp_bench_0256_same    122524000     6.944ns     0.006ns     6.937ns     6.951ns       0.000       0.000        94.7        27.1
iobuf.cmp_bench_0512_same     88760000     9.258ns     0.016ns     9.241ns     9.280ns       0.000       0.000       136.7        36.5
iobuf.cmp_bench_1024_same     59520000    13.224ns     0.004ns    13.218ns    13.228ns       0.000       0.000       204.7        52.6
iobuf.cmp_bench_2048_same     41128000    17.253ns     0.002ns    17.247ns    17.255ns       0.000       0.000       272.7        68.9
```

rockwotj commented on src/v/bytes/iobuf.cc:
```diff
@@ -106,8 +106,12 @@ bool iobuf::operator==(const iobuf& o) const {
-    // We know these have the same amount of bytes in them, but they might be
-    // chunked differently.
+    if (_frags.is_single_fragment() && o._frags.is_single_fragment()) {
+        return std::string_view{_frags.front()}
```
TBH I haven't checked `==` but I assume it's the same as `<=>`. Let me add more benchmarks

rockwotj commented on src/v/bytes/iobuf.cc:
```diff
@@ -143,6 +147,23 @@ bool iobuf::operator<(const iobuf& o) const {
+        auto n = std::min(lhs.size(), rhs.size());
+        constexpr size_t max_byte_for_byte_size = 32;
+        if (n > max_byte_for_byte_size) {
+            return lhs <=> rhs;
```
SG

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--41e39b7a-3fbd-4787-993f-6ed3d924bf7f-->
<!--TEST RESULTS-->
<!--62a4b5ac-b07b-4918-93eb-e233573150d6-->
<!--TEST RESULTS-->
<!--0950ffbb-94d3-47d8-9690-6da8450c2216-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74265">build#74265</a></summary>


| test_class                                            | test_method                                           | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:------------------------------------------------------|:------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/datalake/tests/gtest_record_multiplexer_test    | src/v/datalake/tests/gtest_record_multiplexer_test    |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74265#0199e8ac-a1e9-4e92-b7a5-c2954ea3aafe | FAIL          | 0/1      |          |                |
| src/v/serde/parquet/tests/column_stats_collector_test | src/v/serde/parquet/tests/column_stats_collector_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74265#0199e8ac-a1e9-4e92-b7a5-c2954ea3aafe | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74562">build#74562</a></summary>


| test_class                                            | test_method                                           | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:------------------------------------------------------|:------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest                  | test_continuous_group_sync                            | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74562#019a03c4-fa7e-4bd3-b967-ae9c08631f94 | FLAKY         | 19/21    | upstream reliability is '99.77324263038548'. current run reliability is '90.47619047619048'. drift is 9.29705 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| WriteCachingFailureInjectionTest                      | test_unavoidable_data_loss                            | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74562#019a03c4-fa84-4ecc-af44-d4223036f1c6 | FLAKY         | 18/21    | upstream reliability is '91.96102314250913'. current run reliability is '85.71428571428571'. drift is 6.24674 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss     |
| src/v/datalake/tests/gtest_record_multiplexer_test    | src/v/datalake/tests/gtest_record_multiplexer_test    |                                                                                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/74562#019a037f-c93a-45e8-bb14-21f6c88fa75d | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                               |
| src/v/serde/parquet/tests/column_stats_collector_test | src/v/serde/parquet/tests/column_stats_collector_test |                                                                                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/74562#019a037f-c93a-45e8-bb14-21f6c88fa75d | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74610">build#74610</a></summary>


| test_class         | test_method    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     | test_history                                                                                                                                    |
|:-------------------|:---------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|
| QuickTerminateTest | test_terminate | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74610#019a0788-3a0b-4621-b233-d80a49652a57 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate |
</details>
<!--/body-->

rockwotj commented:
Latest push:

```
test                        iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.cmp_bench_0000         335662000     2.674ns     0.001ns     2.665ns     2.678ns       0.000       0.000        39.2        10.2
iobuf.cmp_bench_0001         149166000     6.376ns     0.004ns     6.371ns     6.414ns       0.000       0.000       100.7        25.7
iobuf.cmp_bench_0002         313382000     2.853ns     0.000ns     2.850ns     2.856ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0003         312722000     2.856ns     0.001ns     2.852ns     2.857ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0004         313198000     2.856ns     0.001ns     2.854ns     2.863ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0016         305130000     2.860ns     0.005ns     2.853ns     2.881ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0064         293590000     2.856ns     0.001ns     2.849ns     2.857ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0128         276720000     2.858ns     0.002ns     2.856ns     2.863ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0256         242960000     2.858ns     0.001ns     2.857ns     2.861ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0512         202758000     2.856ns     0.001ns     2.851ns     2.857ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_1024         146530000     2.856ns     0.002ns     2.854ns     2.858ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_2048          94908000     2.851ns     0.003ns     2.835ns     2.854ns       0.000       0.000        44.7        10.9
iobuf.cmp_bench_0000_same    339454000     2.672ns     0.002ns     2.667ns     2.678ns       0.000       0.000        39.2        10.2
iobuf.cmp_bench_0001_same    149274000     6.371ns     0.002ns     6.361ns     6.374ns       0.000       0.000       100.7        25.7
iobuf.cmp_bench_0002_same    119576000     8.035ns     0.005ns     8.024ns     8.040ns       0.000       0.000       115.7        32.7
iobuf.cmp_bench_0003_same    116014000     8.277ns     0.002ns     8.275ns     8.281ns       0.000       0.000       121.7        33.7
iobuf.cmp_bench_0004_same    116052000     8.274ns     0.001ns     8.267ns     8.277ns       0.000       0.000       127.7        33.7
iobuf.cmp_bench_0016_same    115200000     8.274ns     0.003ns     8.271ns     8.286ns       0.000       0.000       127.7        33.7
iobuf.cmp_bench_0064_same    113270000     8.278ns     0.006ns     8.263ns     8.284ns       0.000       0.000       125.7        33.7
iobuf.cmp_bench_0128_same    103446000     8.918ns     0.004ns     8.900ns     8.922ns       0.000       0.000       137.7        36.4
iobuf.cmp_bench_0256_same     88876000    10.090ns     0.040ns    10.050ns    10.155ns       0.000       0.000       155.7        41.3
iobuf.cmp_bench_0512_same     71108000    12.134ns     0.006ns    12.122ns    12.145ns       0.000       0.000       197.7        49.7
iobuf.cmp_bench_1024_same     54258000    15.015ns     0.005ns    15.010ns    15.025ns       0.000       0.000       251.7        61.8
iobuf.cmp_bench_2048_same     38422000    18.657ns     0.007ns    18.636ns    18.664ns       0.000       0.000       319.7        76.9
iobuf.eq_bench_small         189954000     4.699ns     0.009ns     4.690ns     4.724ns       0.000       0.000        56.7        18.6
iobuf.eq_bench_small_same    162578000     5.586ns     0.005ns     5.573ns     5.591ns       0.000       0.000        71.7        22.3
iobuf.eq_bench_medium           838000     3.240ns     0.007ns     3.220ns     3.262ns       0.000       0.000        48.7        12.0
iobuf.eq_bench_medium_same      144000     5.909us     2.435ns     5.899us     5.912us       0.000       0.000     41346.2     24792.2
iobuf.eq_bench_large            262000     3.261ns     0.010ns     3.250ns     3.275ns       0.000       0.000        48.7        12.0
iobuf.eq_bench_large_same        46000    18.575us    13.006ns    18.562us    18.614us       0.000       0.000    132176.2     78049.
```

I did not do the full plot to figure out where the best cutoff was, but this seemed like decent results, although all these results seem to skew towards the no shared prefix case.

rockwotj committed:
iobuf: always compare prefixes

This means no matter how we layout the iobufs we can get a fast path for
non-shared prefixes.

rockwotj committed:
iobuf: prevent UB in <=>

rockwotj commented:
Added a commit to prevent UB

rockwotj commented:
/microbench

vbotbuildovich commented:
<!--caed638c-500a-4efc-bd27-b3771f00e4ed-->
<!--75e1316d-1a7f-47c6-8e13-411a8aa8562d-->
<!--body-->
Performance change detected in https://buildkite.com/redpanda/redpanda/builds/74792#019a1172-c469-4de2-9205-5d7a52756fef:
```
Performance changes detected in 31 tests
wasm_transform_rpbench.WasmBenchTest_BatchSize1_RecordSize1_KiB.IdentityTransform: inst -> +0.12pct
crypto_bench_rpbench.openssl_perf_test.hmac_sha256_1k: inst -> +0.19pct
crypto_bench_rpbench.openssl_perf_test.hmac_sha512_1k: inst -> +0.06pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_linear_31_fields: inst -> -0.54pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_linear_31_fields_zstd: inst -> -0.53pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_linear_62_fields: inst -> -0.61pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_linear_62_fields_zstd: inst -> -0.60pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_nested_31_levels: inst -> -0.32pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_nested_31_levels_zstd: inst -> -0.31pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_nested_62_levels: inst -> -0.33pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.avro_385_byte_message_nested_62_levels_zstd: inst -> -0.32pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_381_byte_message_linear_40_fields: inst -> -0.65pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_381_byte_message_linear_40_fields_zstd: inst -> -0.64pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_381_byte_message_linear_80_fields: inst -> -0.94pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_381_byte_message_linear_80_fields_zstd: inst -> -0.58pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_384_byte_message_nested_24_levels: inst -> -0.30pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_384_byte_message_nested_24_levels_zstd: inst -> -0.27pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_386_byte_message_nested_31_levels: inst -> -0.30pct
record_multiplexer_rpbench.record_multiplexer_bench_fixture.protobuf_386_byte_message_nested_31_levels_zstd: inst -> -0.29pct
iobuf_rpbench.iobuf.append_bench_large: inst -> -0.12pct
iobuf_rpbench.iobuf.append_bench_small: inst -> -0.02pct
iobuf_rpbench.iobuf.eq_bench_large: inst -> -99.94pct
iobuf_rpbench.iobuf.eq_bench_medium: inst -> -99.81pct
iobuf_rpbench.iobuf.eq_bench_small: inst -> -48.91pct
column_writer_rpbench.column_writer_bench.byte_array_type_400_byte_compressed: inst -> -0.34pct
column_writer_rpbench.column_writer_bench.byte_array_type_400_byte_uncompressed: inst -> -1.63pct
column_writer_rpbench.column_writer_bench.byte_array_type_40_byte_compressed: inst -> -0.92pct
column_writer_rpbench.column_writer_bench.byte_array_type_40_byte_uncompressed: inst -> -4.28pct
column_writer_rpbench.column_writer_bench.byte_array_type_4_byte_compressed: inst -> -3.56pct
column_writer_rpbench.column_writer_bench.byte_array_type_4_byte_uncompressed: allocs -> -0.06pct
column_writer_rpbench.column_writer_bench.byte_array_type_4_byte_uncompressed: inst -> -5.87pct
```
See https://redpandadata.atlassian.net/wiki/x/LQAqLg for docs
<!--/body-->

rockwotj commented:
this claws back some of the "losses". We can optimize more in another change if desired
